# 📅 Day 4 of 100 — Summary Functions & Mapping

📂 **GitHub Repo:** [100-Days-of-Data-Science / Day 004](https://github.com/imshoaibaslam/100-Days-of-Data-Science/tree/main/Day-04-Datascience)

---

## 🗂️ Agenda
1. Summary Functions — `describe()`, `mean()`, `unique()`, `value_counts()`
2. Mapping with `map()`
3. Mapping with `apply()`
4. Vectorized Operations — The Fastest Way

---

## 🎯 Deliverables
- [ ] Use `describe()` on numeric and text columns
- [ ] Calculate mean, unique values, and frequency counts
- [ ] Transform a column using `map()` and a lambda
- [ ] Transform rows using `apply()` with a custom function
- [ ] Perform the same transformation using vectorized operations

---

## 📚 Learning Topics

| # | Topic | Description |
|---|-------|-------------|
| 1 | `describe()` | High-level summary of a column |
| 2 | `mean()` / `unique()` / `value_counts()` | Focused single-stat functions |
| 3 | `map()` | Transform values in a single column |
| 4 | `apply()` | Apply custom logic row by row or column by column |
| 5 | Vectorized Ops | Fastest way to transform — no loops needed |





## 🔧 Imports



In [1]:
import pandas as pd
import numpy as np

---

## Step 0 — Load the Dataset

In [2]:
# Load the Wine Reviews dataset
reviews = pd.read_csv(r"C:\Users\aslam\Downloads\winemag-data-130k-v2.csv\winemag-data-130k-v2.csv", index_col=0)
reviews.head(3)

,country,description,designation,points,price,province,region_1,region_2,taster_name,taster_twitter_handle,title,variety,winery
0,Italy,"Aromas include tropical fruit, broom, brimston...",Vulkà Bianco,87,NaN,Sicily & Sardinia,Etna,NaN,Kerin O’Keefe,@kerinokeefe,Nicosia 2013 Vulkà Bianco (Etna),White Blend,Nicosia
1,Portugal,"This is ripe and fruity, a wine that is smooth...",Avidagos,87,15.0,Douro,NaN,NaN,Roger Voss,@vossroger,Quinta dos Avidagos 2011 Avidagos Red (Douro),Portuguese Red,Quinta dos Avidagos
2,US,"Tart and snappy, the flavors of lime flesh and...",NaN,87,14.0,Oregon,Willamette Valley,Willamette Valley,Paul Gregutt,@paulgwine,Rainstorm 2013 Pinot Gris (Willamette Valley),Pinot Gris,Rainstorm


---

## Step 1 — Summary Functions

Summary functions give us a **quick health report** of our data.
Instead of reading 130,000 rows, we extract key statistics instantly.

#### `describe()` — On a Numeric Column

Think of this as a **one-line report** for any numeric column.

| Stat | Meaning |
|------|---------|
| `count` | How many non-null values exist |
| `mean` | The average value |
| `std` | How spread out the values are |
| `min` | Smallest value |
| `25%` | 25% of values fall below this |
| `50%` | The middle value (median) |
| `75%` | 75% of values fall below this |
| `max` | Largest value |

In [3]:
# Summary of the numeric points column
reviews.points.describe()

count    129971.000000
mean         88.447138
std           3.039730
min          80.000000
25%          86.000000
50%          88.000000
75%          91.000000
max         100.000000
Name: points, dtype: float64

#### `describe()` — On a Text Column

`describe()` is **type-aware** — it behaves differently for text vs numbers.

For text columns it returns:
- `count` → non-null values
- `unique` → how many distinct values exist
- `top` → the most frequent value
- `freq` → how many times top appears

In [4]:
# Summary of the text taster_name column
reviews.taster_name.describe()

count         103727
unique            19
top       Roger Voss
freq           25514
Name: taster_name, dtype: object

#### `mean()` — Average Value

When you only need **one specific statistic**, use focused functions instead of `describe()`.

In [6]:
# What is the average wine score?
reviews.points.mean()

np.float64(88.44713820775404)

#### `unique()` — All Distinct Values

Use this when you want to know **what categories exist** in a column.

In [7]:
# Which tasters exist in this dataset?
reviews.taster_name.unique()

array(['Kerin O’Keefe', 'Roger Voss', 'Paul Gregutt',
       'Alexander Peartree', 'Michael Schachner', 'Anna Lee C. Iijima',
       'Virginie Boone', 'Matt Kettmann', nan, 'Sean P. Sullivan',
       'Jim Gordon', 'Joe Czerwinski', 'Anne Krebiehl\xa0MW',
       'Lauren Buzzeo', 'Mike DeSimone', 'Jeff Jenssen',
       'Susan Kostrzewa', 'Carrie Dykes', 'Fiona Adams',
       'Christina Pickard'], dtype=object)

#### `value_counts()` — Frequency of Each Value

One of the **most used functions in real data work.**
It tells you each unique value AND how many times it appears.

In [8]:
# Which taster reviewed the most wines?
reviews.taster_name.value_counts()

taster_name
Roger Voss            25514
Michael Schachner     15134
Kerin O’Keefe         10776
Virginie Boone         9537
Paul Gregutt           9532
Matt Kettmann          6332
Joe Czerwinski         5147
Sean P. Sullivan       4966
Anna Lee C. Iijima     4415
Jim Gordon             4177
Anne Krebiehl MW       3685
Lauren Buzzeo          1835
Susan Kostrzewa        1085
Mike DeSimone           514
Jeff Jenssen            491
Alexander Peartree      415
Carrie Dykes            139
Fiona Adams              27
Christina Pickard         6
Name: count, dtype: int64

---

## Step 2 — Mapping with `map()`

### 🗺 What is Mapping?

Mapping means:
> Take each value → apply a rule → produce a new value

Example:
```
Old Value    →    New Value
87           →    -1.44  (below average)
90           →    +1.55  (above average)
```

This is called **remeaning** — centering the data around 0.

### What is `lambda`?

A `lambda` is just a quick one-line function.

Instead of writing:
```python
def subtract_mean(p):
    return p - mean
```
We write:
```python
lambda p: p - mean
```
Same thing. Just shorter.

In [9]:
# Step 1 — Calculate the mean
review_points_mean = reviews.points.mean()
print("Average score:", review_points_mean)

Average score: 88.44713820775404


In [10]:
# Step 2 — Subtract mean from every value using map()
# For each point value p → return p minus the mean
reviews.points.map(lambda p: p - review_points_mean)

0        -1.447138
1        -1.447138
2        -1.447138
3        -1.447138
4        -1.447138
            ...   
129966    1.552862
129967    1.552862
129968    1.552862
129969    1.552862
129970    1.552862
Name: points, Length: 129971, dtype: float64

**What just happened?**

- `map()` went through every single row in `points`
- For each value, it subtracted the average
- Negative result → that wine scored **below average**
- Positive result → that wine scored **above average**
- `map()` returns a **new Series** — the original data is untouched

---

## Step 3 — Mapping with `apply()`

### What is `apply()`?

- `map()` works on **one column** (a Series)
- `apply()` works on the **entire DataFrame** — row by row or column by column

### What is a Function?

A function is a **reusable block of code** that takes an input, does something, and returns an output.

```
Input → Function → Output
  87  →  remean  →  -1.44
```

Structure:
```python
def function_name(parameter):
    # do something
    return result
```

### `axis='columns'` vs `axis='index'`

| axis | Meaning | Direction |
|------|---------|----------|
| `axis='columns'` | Apply function **row by row** | → horizontal |
| `axis='index'` | Apply function **column by column** | ↓ vertical |

In [11]:
# Define a function that takes a row and adjusts its points
def remean_points(row):
    row.points = row.points - review_points_mean
    return row

# Apply it row by row across the entire DataFrame
reviews.apply(remean_points, axis='columns')

,country,description,designation,points,price,province,region_1,region_2,taster_name,taster_twitter_handle,title,variety,winery
0,Italy,"Aromas include tropical fruit, broom, brimston...",Vulkà Bianco,-1.447138,NaN,Sicily & Sardinia,Etna,NaN,Kerin O’Keefe,@kerinokeefe,Nicosia 2013 Vulkà Bianco (Etna),White Blend,Nicosia
1,Portugal,"This is ripe and fruity, a wine that is smooth...",Avidagos,-1.447138,15.0,Douro,NaN,NaN,Roger Voss,@vossroger,Quinta dos Avidagos 2011 Avidagos Red (Douro),Portuguese Red,Quinta dos Avidagos
2,US,"Tart and snappy, the flavors of lime flesh and...",NaN,-1.447138,14.0,Oregon,Willamette Valley,Willamette Valley,Paul Gregutt,@paulgwine,Rainstorm 2013 Pinot Gris (Willamette Valley),Pinot Gris,Rainstorm
3,US,"Pineapple rind, lemon pith and orange blossom ...",Reserve Late Harvest,-1.447138,13.0,Michigan,Lake Michigan Shore,NaN,Alexander Peartree,NaN,St. Julian 2013 Reserve Late Harvest Riesling ...,Riesling,St. Julian
4,US,"Much like the regular bottling from 2012, this...",Vintner's Reserve Wild Child Block,-1.447138,65.0,Oregon,Willamette Valley,Willamette Valley,Paul Gregutt,@paulgwine,Sweet Cheeks 2012 Vintner's Reserve Wild Child...,Pinot Noir,Sweet Cheeks
...,...,...,...,...,...,...,...,...,...,...,...,...,...
129966,Germany,Notes of honeysuckle and cantaloupe sweeten th...,Brauneberger Juffer-Sonnenuhr Spätlese,1.552862,28.0,Mosel,NaN,NaN,Anna Lee C. Iijima,NaN,Dr. H. Thanisch (Erben Müller-Burggraef) 2013 ...,Riesling,Dr. H. Thanisch (Erben Müller-Burggraef)
129967,US,Citation is given as much as a decade of bottl...,NaN,1.552862,75.0,Oregon,Oregon,Oregon Other,Paul Gregutt,@paulgwine,Citation 2004 Pinot Noir (Oregon),Pinot Noir,Citation
129968,France,Well-drained gravel soil gives this wine its c...,Kritt,1.552862,30.0,Alsace,Alsace,NaN,Roger Voss,@vossroger,Domaine Gresser 2013 Kritt Gewurztraminer (Als...,Gewürztraminer,Domaine Gresser
129969,France,"A dry style of Pinot Gris, this is crisp with ...",NaN,1.552862,32.0,Alsace,Alsace,NaN,Roger Voss,@vossroger,Domaine Marcel Deiss 2012 Pinot Gris (Alsace),Pinot Gris,Domaine Marcel Deiss


**What just happened?**

- `apply()` passed **one row at a time** into `remean_points()`
- Inside the function, we accessed `row.points` and subtracted the mean
- The modified row was returned
- This repeated for all 129,971 rows

⚠️ **Important:** `apply()` does NOT modify the original DataFrame.
It returns a new transformed copy.

In [12]:
# Proof — original data is unchanged
reviews.head(1)

,country,description,designation,points,price,province,region_1,region_2,taster_name,taster_twitter_handle,title,variety,winery
0,Italy,"Aromas include tropical fruit, broom, brimston...",Vulkà Bianco,87,NaN,Sicily & Sardinia,Etna,NaN,Kerin O’Keefe,@kerinokeefe,Nicosia 2013 Vulkà Bianco (Etna),White Blend,Nicosia


---

## Step 4 — Vectorized Operations (The Fastest Way)

### What is Vectorization?

Instead of going **value by value**, pandas applies the operation to the **entire column at once**.

No loops. No `map()`. No `apply()`.
Just direct math on the column.

This is **much faster** — especially on large datasets like 130,000 rows.

In [13]:
# Same result as map() — but faster and cleaner
reviews.points - review_points_mean

0        -1.447138
1        -1.447138
2        -1.447138
3        -1.447138
4        -1.447138
            ...   
129966    1.552862
129967    1.552862
129968    1.552862
129969    1.552862
129970    1.552862
Name: points, Length: 129971, dtype: float64

**What pandas does behind the scenes:**

- Left side → entire column (many values)
- Right side → single mean value
- pandas subtracts the mean from **every value automatically**

No function needed. No lambda needed. Just math.

#### Combining Two Columns (Vectorized)

In [14]:
# Combine country and region into one column
reviews.country + " - " + reviews.region_1

0                     Italy - Etna
1                              NaN
2           US - Willamette Valley
3         US - Lake Michigan Shore
4           US - Willamette Valley
                    ...           
129966                         NaN
129967                 US - Oregon
129968             France - Alsace
129969             France - Alsace
129970             France - Alsace
Length: 129971, dtype: object

**Result:** Each row combines like this:
```
Italy + " - " + Etna  →  Italy - Etna
```

If `region_1` is `NaN`, the whole result becomes `NaN` — pandas handles this automatically.

---

## 💡 When to Use What

| Situation | Use |
|-----------|-----|
| Quick overview of a column | `describe()` |
| Single statistic needed | `mean()`, `min()`, `max()` |
| What categories exist | `unique()` |
| Frequency of each value | `value_counts()` |
| Transform one column value by value | `map()` |
| Complex logic across multiple columns | `apply()` |
| Simple math between columns | Vectorized operators |

---

## 🌍 Real World Usage

**`describe()` + `value_counts()`**
The first two things a data scientist runs after loading any new dataset. They instantly reveal the shape, distribution, and quality of the data.

**`map()`**
Used to create transformed columns — price categories, score buckets, label encoding for machine learning features.

**`apply()`**
Used when logic requires multiple columns together — for example, calculating a value score from points divided by price.

**Vectorized Operations**
The industry standard for production code. Faster, cleaner, and the approach all senior data scientists default to.

---

## 🔑 Key Takeaways

- `describe()` is **type-aware** — behaves differently for numbers vs text
- `value_counts()` is one of the **most used** functions in real data work
- `map()` works on a **single column** only
- `apply()` works on **entire rows or columns** of a DataFrame
- Neither `map()` nor `apply()` modifies the original data — they return **new copies**
- **Vectorized operations** are always faster — prefer them over `map()` and `apply()` when possible
- `axis='columns'` → row by row | `axis='index'` → column by column

---

## 🏁 Conclusion

Today we covered **two of the most essential skills in pandas** — summarizing data and transforming it.

Summary functions let you instantly understand what's in your dataset without reading every row. Mapping and vectorized operations let you reshape and transform that data into exactly the format your analysis needs.

These tools are used in **every single real-world data science project** — from exploratory analysis to feature engineering for machine learning. From Day 6 onwards, we will use all of these constantly.

---

## ✅ Day 4 Complete — 96 days to go!

<p align="center">
  <a href="https://github.com/imshoaibaslam/100-Days-of-Data-Science/tree/main/Day-04-Datascience">📂 View Full Code on GitHub</a>
</p>